File: go/0377M-组合总和Ⅳ.go
/*
 * @Author: SukiEva
 * @Date: 2021-12-01 14:52:17
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

func combinationSum4(nums []int, target int) int {
	dp := make([]int, target+1)
	dp[0] = 1
	for j := 0; j <= target; j++ {
		for _, num := range nums {
			if j >= num {
				dp[j] += dp[j-num]
			}
		}
	}
	return dp[target]
}

File: cpp/0377M-组合总和Ⅳ.cpp
/*
 * @Author: SukiEva
 * @Date: 2022-01-11 11:29:25
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int combinationSum4(vector<int>& nums, int target) {
        vector<int> dp(target + 1);
        dp[0] = 1;
        for (int j = 0; j <= target; j++) {
            for (int num : nums) {
                if (j - num >= 0 &&
                    dp[j] < INT_MAX - dp[j - num]) {  // 测试中有数据相加溢出
                    dp[j] += dp[j - num];
                }
            }
        }
        return dp[target];
    }
};